### PR TITLE
Setting up the authentication system

### LINKED ISSUE #36: Authentication and user roles

### LINKED ISSUE #43: Use cookies to persist the user's JWT token

### PR DESCRIPTION
**Pull request by gzcharleszhang**
closes #43 closes #36 
### Workflow
- A token encodes the user id and expiration time (set to 14 days for now).
- Whenever a user logs in, we issue them a jwt token and store it in a cookie.
- On every request, we extract the token's expiration time and user id.
- We throw a 401 if the token has expired. The user needs to login again.
- We refresh the token if it's going to expire in 3 days (refresh period).

### Example
- Steven XU has an account with us and wants to login, we give him a jwt token that expires in 14 days along with his user id `MR_CALC_AWARD`.
- This token is then stored inside a cookie in Steven's browser.
- **A month later**, Steven wants to add a new CourseRecord, he issues a request to the server **without** needing to log in.
- The server then decodes the token and extracts the token's expiration time.
- Since the token was issued over 30 days ago, thus the server rejects this request and Steven is forced to login again.
- Steven receives a new token by repeating the above steps.
- After 13 days, Steven wants to add a new CourseRecord, he issues a request to the server **without** needing to log in.
- The server then decodes the token and extracts the token's expiration time.
- Since the token is not expired, we allow the request through and move onto the role + permission validation stage.
- Since the token is within the refresh period (within 3 days of expiring), we issue Steven a new token that will expire in another 14 days.

### Workflow
- Each `User` has a `Role`
- Each `Role` has a certain defined permission level access
- We can specify the permission level required for each route through middlewares
- Requests to protected routes are rejected if the role in that request does not have the permission level access defined by the route

### Example
- Steven XU signed up for an account with us with the role `Conrad`.
- Steven tries to access a service that requires admin permission level access.
- The server decodes Steven's cookie and extract his user id
- The server then fetches Steven's role using his user id.
- Since the `Conrad` role does not have admin permission level access, Steven's request is rejected.

### CODE REVIEW

File: internal/app/components/permissions/permissions.go [hunk 1]
```diff
@@ -0,0 +1,9 @@
+package permissions
+
+type Permission string
+
+const (
+	Admin           = "admin"
```
Comment (infinitecold) [inline internal/app/components/permissions/permissions.go]: should these be of type Permission

File: internal/app/components/users/users.go [hunk 2]
```diff
@@ -3,59 +3,81 @@ package users
...
-		Id:        newUserId,
-		FirstName: firstName,
-		LastName:  lastName,
-		Password:  password,
+		Id:               newUserId,
+		FirstName:        firstName,
+		LastName:         lastName,
+		Password:         password,
+		Email:            email,
+		PermissionAccess: permissions.Authenticated, // default to authenticated
```
Comment (infinitecold) [inline internal/app/components/users/users.go]: why is this the default

**Comment by gzcharleszhang** [comment]
@infinitecold seems like permission is kinda confusing, gonna do some refactoring and have both roles and permissions

**Comment by infinitecold** [comment]
sounds good, i'm not familiar with authentication stuff so i can't comment on the design but from a pure code perspective the rest of the PR looks fine

File: internal/app/components/users/users.go [hunk 3]
```diff
@@ -60,12 +60,12 @@ func CreateUser(ctx context.Context, firstName FirstName, lastName LastName,
...
-		LastName:         lastName,
-		Password:         password,
-		Email:            email,
-		PermissionAccess: permissions.Authenticated, // default to authenticated
+		Id:        newUserId,
+		FirstName: firstName,
+		LastName:  lastName,
+		Password:  password,
+		Email:     email,
+		Role:      roles.NewConrad(), // default to conrad
```
Comment (infinitecold) [inline internal/app/components/users/users.go]: this just looks wrong lmao

**Pull request merged**